File: src/main/java/org/soraworld/lime/util/LimitUtils.java
package org.soraworld.lime.util;

import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.soraworld.lime.config.LangKeys;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class LimitUtils {

    private static final SimpleDateFormat DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd-HH:mm");
    private static final Pattern PATTERN = Pattern.compile("[+-]\\d+[mhd]");
    private static final Pattern LIMITIME = Pattern.compile("\\[limitime:([+-]\\d+[mhd])+]");
    private static final Pattern DEADLINE = Pattern.compile("\\[deadline:[0-9-:]+]");
    private static final Pattern DATE = Pattern.compile("[0-9]{4}(-[0-9]{2}){3}:[0-9]{2}");

    private static int parse(String exp) {
        int last = exp.length() - 1;
        int val = Integer.valueOf(exp.substring(1, last));
        if (exp.charAt(0) == '-') val *= -1;
        if (exp.charAt(last) == 'h') val *= 60;
        if (exp.charAt(last) == 'd') val *= 1440;
        return val;
    }

    public static long limitime(String limitime) {
        long time = 0;
        Matcher matcher = PATTERN.matcher(limitime);
        while (matcher.find()) {
            time += parse(matcher.group());
        }
        return time * 60000;
    }

    private static boolean deadline(String deadline) {
        Matcher matcher = DATE.matcher(deadline);
        if (matcher.find()) {
            try {
                if (DATE_FORMAT.parse(matcher.group()).before(new Date())) {
                    return true;
                }
            } catch (Throwable e) {
                ServerUtils.console(LangKeys.format("parseException", deadline));
            }
        }
        return false;
    }

    public static String deadLore(Date date) {
        return "[deadline:" + DATE_FORMAT.format(date) + "]";
    }

    public static String limeLore(String limitime) {
        return "[limitime:" + limitime + "]";
    }

    public static void setLore(Player player, ItemStack item, ItemMeta meta, List<String> lore) {
        meta.setLore(lore);
        item.setItemMeta(meta);
        player.setItemInHand(item);
    }

    public static boolean isDead(ItemStack item) {
        if (item == null || item.getItemMeta() == null) return false;
        List<String> lore = item.getItemMeta().getLore();
        for (int i = 0; lore != null && i < lore.size(); i++) {
            String line = lore.get(i);
            Matcher matcher = DEADLINE.matcher(line);
            if (matcher.find() && deadline(matcher.group())) {
                return true;
            }
        }
        return false;
    }

    public static boolean hasLimitime(ItemStack item) {
        if (item != null && item.getItemMeta() != null) {
            ItemMeta meta = item.getItemMeta();
            List<String> lore = meta.getLore();
            for (int i = 0; lore != null && i < lore.size(); i++) {
                String line = lore.get(i);
                Matcher matcher = LIMITIME.matcher(line);
                if (matcher.find()) {
                    Date date = new Date();
                    date.setTime(date.getTime() + limitime(matcher.group()));
                    lore.set(i, deadLore(date));
                    meta.setLore(lore);
                    item.setItemMeta(meta);
                    return true;
                }
            }
        }
        return false;
    }
}

File: src/main/java/org/soraworld/lime/util/LimitUtils.kt
package org.soraworld.lime.util

import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.ItemMeta
import java.text.SimpleDateFormat
import java.util.*
import java.util.regex.Pattern

object LimitUtils {

    private val DATE_FORMAT = SimpleDateFormat("yyyy-MM-dd-HH:mm")
    private val PATTERN = Pattern.compile("[+-]\\d+[mhd]")
    private val LIMITIME = Pattern.compile("\\[limitime:([+-]\\d+[mhd])+]")
    private val DEADLINE = Pattern.compile("\\[deadline:[0-9-:]+]")
    private val DATE = Pattern.compile("[0-9]{4}(-[0-9]{2}){3}:[0-9]{2}")

    private fun parse(exp: String): Int {
        val last = exp.length - 1
        var `val` = Integer.valueOf(exp.substring(1, last))
        if (exp[0] == '-') `val` *= -1
        if (exp[last] == 'h') `val` *= 60
        if (exp[last] == 'd') `val` *= 1440
        return `val`
    }

    fun limitime(limitime: String): Long {
        var time: Long = 0
        val matcher = PATTERN.matcher(limitime)
        while (matcher.find()) {
            time += parse(matcher.group()).toLong()
        }
        return time * 60000
    }

    private fun deadline(deadline: String): Boolean {
        val matcher = DATE.matcher(deadline)
        if (matcher.find()) {
            try {
                if (DATE_FORMAT.parse(matcher.group()).before(Date())) {
                    return true
                }
            } catch (e: Throwable) {
                println("parseException$deadline")
            }

        }
        return false
    }

    fun deadLore(date: Date): String {
        return "[deadline:" + DATE_FORMAT.format(date) + "]"
    }

    fun limeLore(limitime: String): String {
        return "[limitime:$limitime]"
    }

    fun setLore(player: Player, item: ItemStack, meta: ItemMeta, lore: List<String>) {
        meta.lore = lore
        item.itemMeta = meta
        player.itemInHand = item
    }

    fun isDead(item: ItemStack?): Boolean {
        if (item == null || item.itemMeta == null) return false
        val lore = item.itemMeta.lore
        var i = 0
        while (lore != null && i < lore.size) {
            val line = lore[i]
            val matcher = DEADLINE.matcher(line)
            if (matcher.find() && deadline(matcher.group())) {
                return true
            }
            i++
        }
        return false
    }

    fun hasLimitime(item: ItemStack?): Boolean {
        if (item != null && item.itemMeta != null) {
            val meta = item.itemMeta
            val lore = meta.lore
            var i = 0
            while (lore != null && i < lore.size) {
                val line = lore[i]
                val matcher = LIMITIME.matcher(line)
                if (matcher.find()) {
                    val date = Date()
                    date.time = date.time + limitime(matcher.group())
                    lore[i] = deadLore(date)
                    meta.lore = lore
                    item.itemMeta = meta
                    return true
                }
                i++
            }
        }
        return false
    }
}
